### PR TITLE
GMG Newton: correct Stokes rhs

### PR DESCRIPTION
**Pull request by zjiaqi2018**
Adding the Newton part to the function`StokesMatrixFreeHandlerImplementation::correct_stokes_rhs()`. @tjhei

### CODE REVIEW

**Comment by tjhei** [comment]
Looks reasonable. Does it make a difference for the Newton benchmarks? It could be that it is only a tiny difference in the updates close to the boundary, that doesn't matter...
/rebuild

**Comment by tjhei** [comment]
No difference in test output...

File: source/simulator/stokes_matrix_free.cc [hunk 1]
```diff
@@ -1837,7 +1837,18 @@ namespace aspect
                                                           velocity.get_symmetric_gradient (q);
             VectorizedArray<double> pres = pressure.get_value(q);
             VectorizedArray<double> div = trace(sym_grad_u);
-            pressure.submit_value   (sim.pressure_scaling*div, q);
+
+            if (active_cell_data.enable_newton_derivatives)
+              {
+                // Note that derivative_scaling_factor has already been multiplied to newton_factor_wrt_pressure_table.
+                VectorizedArray<double> newton_pressure_term =
```
Comment (bangerth) [inline source/simulator/stokes_matrix_free.cc:1844]:
```suggestion
                const VectorizedArray<double> newton_pressure_term =
```

File: source/simulator/stokes_matrix_free.cc [hunk 2]
```diff
@@ -1837,7 +1837,18 @@ namespace aspect
                                                           velocity.get_symmetric_gradient (q);
             VectorizedArray<double> pres = pressure.get_value(q);
             VectorizedArray<double> div = trace(sym_grad_u);
```
Comment (bangerth) [inline source/simulator/stokes_matrix_free.cc:1839]: These two variables could be `const` as well, I believe.

As a general rule, mark everything as `const` that can be marked as `const`.

File: source/simulator/stokes_matrix_free.cc [hunk 3]
```diff
@@ -1848,6 +1859,20 @@ namespace aspect
               for (unsigned int d=0; d<dim; ++d)
                 sym_grad_u[d][d] -= viscosity_x_2/3.0*div;
 
+            if (active_cell_data.enable_newton_derivatives)
+              {
+                SymmetricTensor<2,dim,VectorizedArray<double>> grads_phi_u_i = velocity.get_symmetric_gradient(q);
+
+                sym_grad_u +=  (grads_phi_u_i * active_cell_data.strain_rate_table(cell,q))
+                               * active_cell_data.newton_factor_wrt_strain_rate_table(cell,q);
```
Comment (bangerth) [inline source/simulator/stokes_matrix_free.cc:1867]: This is now no longer the symmetric gradient, i.e., the variable *name* no longer matches its *value*. I think it would be better to do
```
  if (enable_newton_derivatives)
  {
    const something = sym_grad_u + ...;
    velocity.submit_sym_grad(...);
  }
  else
    velocity.submit_sym_grad (existing arguments);
```

**Review by bangerth — CHANGES_REQUESTED** [review]

File: source/simulator/stokes_matrix_free.cc (hunk 3 above)
Comment (zjiaqi2018) [reply source/simulator/stokes_matrix_free.cc:1867]: > No difference in test output...

Exactly! Is it because in the Newton method, the rhs is already the residual, so the system already has homogeneous Dirichlet boundary conditions?

**Comment by tjhei** [comment]
You are correct, see https://github.com/geodynamics/aspect/blob/9eaaf4fbbdb5e81e1d2855464b8de45a139b8e97/source/simulator/core.cc#L1286
We never include Newton terms in step 0 and after that we solve with zero boundary conditions.

This means you can replace this PR by a comment stating that we don't need to include the Newton terms because of that.
Sorry.

**Review by tjhei — APPROVED** [review]

**Comment by tjhei** [comment]
@bangerth will be unavailable for a week or so, so I will go ahead and merge.

**Comment by bangerth** [comment]
Thanks, looked good!
